### ISSUE TITLE
[temp.arg.nontype] p3 "subobject thereof" is not clear

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#115]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[temp.arg.nontype] p3
> For a non-type template-parameter of reference or pointer type, or for each non-static data member of reference or pointer type in a non-type template-parameter of class type or **subobject thereof**, the reference or pointer value shall not refer to or be the address of (respectively):

What does "subobject thereof" intend to mean? Does it refer to the subobject of the non-static data member of reference or pointer type, or something else?  Consider this example  
````cpp
struct Test{
    char const* ptr;
};
struct A{
    Test arr[1];
};
struct B{
    A a;
};
template<B b>
struct C{};

C<B{{"abc"}}> c;
````
Clang and GCC reject this example and friendly say
> pointer to subobject of string literal is not allowed in a template argument

[temp.arg.nontype] p3 should reject this example, however, the current wording is not clear that it rejects this complex case.  The non-static data member of class `B` is neither of reference nor pointer type. As well, the subobject of `B::a` is also not of a pointer or reference type, rather, it is of array type.   

The current wording sounds like it is phrased as
> for each non-static data member of reference or pointer type in a non-type template-parameter of class type, or
> the subobject of the non-static data member of reference or pointer type

### Suggested resolution

Change [temp.arg.nontype] p3 to
>  For an object `o`, define the set `S(o)` of subobjects of `o` as follows: 
>> - if `o` is of non-union class type, `S(o)` comprises each non-static data member subobject `x` of `o` and the objects in each set `S(x)`. 
>> - if `o` is of array type, `S(o)` comprises each element subobject `x` of `o` and the objects in each set `S(x)`.
>> - if `o` is of union type, `S(o)` comprises the active variant member subobject `x` of `o` and the objects in set `S(x)`
>> - otherwise, `S(o)` is empty
>
> For a non-type template-parameter of reference or pointer type, or for the non-type template-parameter of class type, each object in `S(o)` that is of reference or pointer type, where `o` is an object of the class type, the reference or pointer value shall not refer to or be the address of (respectively):
>> - [...]

To make [temp.arg.nontype] p3 can **recursively apply** to the non-type template-parameter of class type.

**Comment by jensmaurer** [comment]
Obviously, pointers or references don't have subobjects; only classes can have subobjects.

I'm not seeing how this is unclear.

**Comment by xmh0511** [comment]
> only classes can have subobjects.

As shown in the first example, the rule should also apply to the element of an array that is a subobject of a class object. Moreover, the rule also applies to the subobject of the element of the array of the class object. This is an unclear point, at least, the meaning of "subobject thereof" is not clear. Does it mean the subobject of the class object that is being talked about, or it can touch more subobjects at the deep level?

**Comment by jensmaurer** [comment]
The definition of "subobject" is recursive, I think.
And we're talking about the active member of a union, apparently.

**Comment by xmh0511** [comment]
> The definition of "subobject" is recursive, I think.

This is the subject of https://github.com/cplusplus/draft/issues/2310, however, it is still unclear in the current draft. non-active and active variant members of a union are both called subobjects, I think. At least, we didn't specify which subobjects we are talking about in the current wording.
